lib/codecs/src/encoding/format/csv.rs

```diff
@@ -0,0 +1,183 @@
+        wtr.flush()?;
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use bytes::BytesMut;
+    use chrono::DateTime;
+    use ordered_float::NotNan;
+    use vector_common::btreemap;
+    use vector_core::event::{LogEvent, Value};
+
+    use super::*;
+
+    #[test]
+    fn build_error_on_empty_fields() {
+        let opts = CsvSerializerOptions::default();
+        let config = CsvSerializerConfig::new(opts);
+        let err = config.build().unwrap_err();
+        assert_eq!(err.to_string(), "At least one CSV field must be specified");
+    }
+
+    #[test]
+    fn serialize_fields() {
+        let event = Event::Log(LogEvent::from(btreemap! {
+            "foo" => Value::from("bar"),
+            "int" => Value::from(123),
+            "comma" => Value::from("abc,bcd"),
+            "float" => Value::Float(NotNan::new(3.1415925).unwrap()),
+            "space" => Value::from("sp ace"),
+            "time" => Value::Timestamp(DateTime::parse_from_rfc3339("2023-02-27T15:04:49.363+08:00").unwrap().into()),
+            "quote" => Value::from("the \"quote\" should be escaped"),
+            "bool" => Value::from(true),
+            "other" => Value::from("data"),
+        }));
+
+        let fields = vec![
+            ConfigOwnedTargetPath::try_from("foo".to_string()).unwrap(),
+            ConfigOwnedTargetPath::try_from("int".to_string()).unwrap(),
+            ConfigOwnedTargetPath::try_from("comma".to_string()).unwrap(),
+            ConfigOwnedTargetPath::try_from("float".to_string()).unwrap(),
+            ConfigOwnedTargetPath::try_from("missing".to_string()).unwrap(),
+            ConfigOwnedTargetPath::try_from("space".to_string()).unwrap(),
+            ConfigOwnedTargetPath::try_from("time".to_string()).unwrap(),
+            ConfigOwnedTargetPath::try_from("quote".to_string()).unwrap(),
+            ConfigOwnedTargetPath::try_from("bool".to_string()).unwrap(),
+        ];
+        let config = CsvSerializerConfig::new(CsvSerializerOptions { fields });
+        let mut serializer = config.build().unwrap();
+        let mut bytes = BytesMut::new();
+
+        serializer.encode(event, &mut bytes).unwrap();
+
+        assert_eq!(
+            bytes.freeze(),
+            b"bar,123,\"abc,bcd\",3.1415925,,sp ace,2023-02-27T07:04:49.363+00:00,\"the \"\"quote\"\" should be escaped\",true".as_slice()
+        );
+    }
+
+    #[test]
+    fn serialize_order() {
+        let event = Event::Log(LogEvent::from(btreemap! {
+            "field1" => Value::from("value1"),
+            "field2" => Value::from("value2"),
+            "field3" => Value::from("value3"),
+            "field4" => Value::from("value4"),
+            "field5" => Value::from("value5"),
+        }));
+        let fields = vec![
+            ConfigOwnedTargetPath::try_from("field1".to_string()).unwrap(),
+            ConfigOwnedTargetPath::try_from("field5".to_string()).unwrap(),
+            ConfigOwnedTargetPath::try_from("field5".to_string()).unwrap(),
```

Comment:
Was one of these intended to be `field4`?

Reply:
Hahaha,  actually this is just by intention, for duplicating and ignoring fields.

Reply:
Fair enough! I didn't even think about duplication being supported.